packages/matplotlib/patches/0004-WebAgg-backend.patch

```diff
@@ -0,0 +1,721 @@
+-        fig_id=fignum,
+-        toolitems=core.NavigationToolbar2WebAgg.toolitems,
+-        canvas=figure.canvas,
+-        port=WebAggApplication.port).decode('utf-8')
++    def get_toolbar_image(cls, image):
++        filename = Path(__file__).parent.parent / f"mpl-data/images/{image}.png"
++        png_bytes = filename.read_bytes()
++        return png_bytes
+
+
+ @_Backend.export
+diff --git a/lib/matplotlib/backends/backend_webagg_core.py b/lib/matplotlib/backends/backend_webagg_core.py
+index 4ceac16995..69bcc1a5bc 100644
+--- a/lib/matplotlib/backends/backend_webagg_core.py
++++ b/lib/matplotlib/backends/backend_webagg_core.py
+@@ -8,17 +8,21 @@
+ # - `backend_webagg.py` contains a concrete implementation of a basic
+ #   application, implemented with asyncio.
+
+-import asyncio
+-import datetime
++import base64
+ from io import BytesIO, StringIO
+ import json
+ import logging
++import mimetypes
+ import os
+ from pathlib import Path
+
++from js import alert, document
+ import numpy as np
+ from PIL import Image
+
++from pyodide.ffi.wrappers import (
++    clear_interval, clear_timeout, set_interval, set_timeout)
++
+ from matplotlib import _api, backend_bases, backend_tools
+ from matplotlib.backends import backend_agg
+ from matplotlib.backend_bases import (
+@@ -78,37 +82,27 @@ def _handle_key(key):
+     return key
+
+
+-class TimerTornado(backend_bases.TimerBase):
++class TimerJs(backend_bases.TimerBase):
+     def __init__(self, *args, **kwargs):
+-        self._timer = None
++        self._timer: int | None = None
+         super().__init__(*args, **kwargs)
+
+     def _timer_start(self):
+-        import tornado
+-
+         self._timer_stop()
+         if self._single:
+-            ioloop = tornado.ioloop.IOLoop.instance()
+-            self._timer = ioloop.add_timeout(
+-                datetime.timedelta(milliseconds=self.interval),
+-                self._on_timer)
++            self._timer = set_timeout(self._on_timer, self.interval)
+         else:
+-            self._timer = tornado.ioloop.PeriodicCallback(
+-                self._on_timer,
+-                max(self.interval, 1e-6))
+-            self._timer.start()
++            self._timer = set_interval(self._on_timer, self.interval)
+
+     def _timer_stop(self):
+-        import tornado
+-
+         if self._timer is None:
+             return
+         elif self._single:
+-            ioloop = tornado.ioloop.IOLoop.instance()
+-            ioloop.remove_timeout(self._timer)
++            clear_timeout(self._timer)
++            self._timer = None
+         else:
+-            self._timer.stop()
+-        self._timer = None
++            clear_interval(self._timer)
++            self._timer = None
+
+     def _timer_set_interval(self):
+         # Only stop and restart it if the timer has already been started
+@@ -117,44 +111,9 @@ class TimerTornado(backend_bases.TimerBase):
+             self._timer_start()
+
+
+-class TimerAsyncio(backend_bases.TimerBase):
```

Comment:
We could also just use `TimerAsyncio` -- I think it would work fine and require a slightly smaller patch. Or is there something I'm misunderstanding?

Reply:
`TimerAsyncio` works, thanks.